prereise/gather/griddata/hifld/data_access/load.py

```diff
@@ -186,6 +189,102 @@ def get_hifld_electric_substations(path):
+                continue
+            # Extract latitude and longitude from the appropriate end of each segment
+            lon1, lat1 = segments[seg1][0 if which1 == "start" else -1]
+            lon2, lat2 = segments[seg2][0 if which2 == "start" else -1]
+            edge_angle = angular_distance(
+                latlon_to_xyz(lat1, lon1), latlon_to_xyz(lat2, lon2)
+            )
+            g.add_edge((seg1, which1), (seg2, which2), weight=edge_angle)
+        # Identify edges to connect the segments
+        tree = nx.algorithms.tree.minimum_spanning_tree(g)
+        # Ensure that the minimum spanning tree is linear
+        if max(dict(nx.degree(tree)).values()) > 2:
+            raise ValueError("The minimum spanning tree isn't linear")
+        return tree
+
+    if len(segments) == 1:
+        # There's only one segment in the list, so no joining is necessary
+        return segments[0]
+
+    segments_df = pd.Series(segments).to_frame(name="segments")
+    segments_df["distance"] = segments_df["segments"].map(
+        lambda x: angular_distance(
+            latlon_to_xyz(x[0][1], x[0][0]),  # first point
+            latlon_to_xyz(x[-1][1], x[-1][0]),  # last point
+        )
+    )
+    filtered_segments = segments_df.loc[segments_df["distance"] > threshold]
+    if len(filtered_segments) == 1:
+        # There's only one segment left in the list, so no joining is necessary
+        return filtered_segments["segments"].iloc[0]
+
+    # Sort the segments by length, in case the minimum spanning tree isn't linear
+    sorted_segments = filtered_segments.sort_values("distance")
+    # Make a minimum spanning tree, dropping small segments as necessary for linearity
+    for i in range(len(sorted_segments)):
+        try:
+            linear_segments = sorted_segments.iloc[i:]["segments"].tolist()
+            tree = _generate_linear_spanning_tree(linear_segments)
+            # We found a good minimum spanning tree, so we can exit
+            break
+        except ValueError:
+            # The minimum spanning tree wasn't good, go to the next iteration
+            pass
+
+    # Identify an endpoint of the minimum spanning tree
+    node_degrees = dict(nx.degree(tree))
+    endpoint = min(k for k, v in node_degrees.items() if v == 1)
+    last_endpoint = None
+    # Traverse the tree, adding segments in order
+    joined_segments = []
+    while True:
+        next_endpoints = list(set(tree[endpoint]) - {last_endpoint})
+        assert len(next_endpoints) == 1
+        next_endpoint = next_endpoints[0]
+        (seg1, which1), (seg2, _) = endpoint, next_endpoints[0]
+        if seg1 == seg2:
+            if which1 == "start":
+                joined_segments += linear_segments[seg1]
+            else:
+                joined_segments += linear_segments[seg1][::-1]
+        if node_degrees[next_endpoint] == 1:
+            break
+        last_endpoint = endpoint
+        endpoint = next_endpoint
```

Comment:
Not sure whether we can use [dfs_preorder_nodes](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.traversal.depth_first_search.dfs_preorder_nodes.html#networkx.algorithms.traversal.depth_first_search.dfs_preorder_nodes) (should be equivalent with `dfs_postorder_nodes` given it is a path graph) here

Reply:
TIL! That should definitely simplify things.

Reply:
I ended up using [df_edges](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.traversal.depth_first_search.dfs_edges.html) instead, which I think should be equivalent, and verified that the behavior is the same as what I had previously.